sim/sample.py

```diff
@@ -0,0 +1,158 @@
+from utils import run
+import math 
+
+import os
+
+class Sample:
+    per_base_error_rate = "0"
+    num_read_pairs = 150000
+
+    def simulate_genome(self):
+        raise NotImplementedError()
+
+    @property
+    def name(self):
+        # TODO: this is ugly, make this human-readable
+        return f"unnamed-{str(id(self))}"
+
+    def simulate_reads(self, simulated_genome_path, simulated_reads_path):
+        # # TODO: explicitly path fasta path to simulate
+        fasta_path = simulated_genome_path + self.name + '.simulated.simseq.genome.fa'
+        simulate_reads(
+            fasta_path, 
+            simulated_reads_path, 
+            sample_name=self.name, 
+            per_base_error_rate=self.per_base_error_rate,
+            num_read_pairs=self.num_read_pairs
+        )
+
+class VcfSample(Sample):
+    def __init__(self, 
+        predef_snp_path, 
+        seed=1, 
+        per_base_error_rate="0",
+        num_read_pairs = 10,
+    ):
+        self.seed = seed
+        self.per_base_error_rate = per_base_error_rate
+        self.num_read_pairs = math.ceil(num_read_pairs)
+
+        # TODO: check predef_snp_path exists
+        self.predef_snp_path = predef_snp_path
+
+    @property
+    def name(self):
+        # extract filename
+        #TODO: this will fail if the extension is not 3 charachters long
+        #TODO: this pattern is getting somewhat unweildy.
+        #   etiher use something more general or save metadata separately
+        basename = os.path.basename(self.predef_snp_path)[:-4]
+        return f"{type(self).__name__}-{basename}-seed{self.seed}-error{self.per_base_error_rate}-readpairs-{self.num_read_pairs}"
+
+    def simulate_genome(self, reference_path, simulated_genome_path):
+        """ Simulate a genome fasta using simuG."""
+        simulate_genome_from_vcf(reference_path, simulated_genome_path, self.predef_snp_path, seed=1)
+
+class RandomSample(Sample):
+    def __init__(self, num_snps=16000, seed=1, per_base_error_rate="0"):
+        self.num_snps = num_snps
+        self.seed = seed
+        self.per_base_error_rate = per_base_error_rate
+    
+    @property
+    def name(self):
+        return f"{type(self).__name__}-snps{self.num_snps}-seed{self.seed}"
+
+    def simulate_genome(self, reference_path, simulated_genome_path):
+        simulate_genome_random_snps(reference_path, simulated_genome_path, num_snps=self.num_snps, seed=self.seed)
+
+# TODO: move these functions inside of the classes
+
+def simulate_genome_random_snps(reference_path, simulated_genome_path, num_snps=16000, seed=1):
```

Comment:
I think these functions should be inside the classes

Reply:
I agree, see the TODO on line 69